collect_app/src/main/java/org/odk/collect/android/formmanagement/FormMediaDownloader.kt

```diff
@@ -0,0 +1,63 @@
+package org.odk.collect.android.formmanagement
+
+import org.odk.collect.android.utilities.FileUtils.copyFile
+import org.odk.collect.android.utilities.FileUtils.interuptablyWriteFile
+import org.odk.collect.async.OngoingWorkListener
+import org.odk.collect.forms.Form
+import org.odk.collect.forms.FormSource
+import org.odk.collect.forms.FormSourceException
+import org.odk.collect.forms.FormsRepository
+import org.odk.collect.forms.MediaFile
+import org.odk.collect.shared.strings.Md5.getMd5Hash
+import java.io.File
+import java.io.IOException
+
+class FormMediaDownloader(
+    private val formsRepository: FormsRepository,
+    private val formSource: FormSource
+) {
+
+    @Throws(IOException::class, FormSourceException::class, InterruptedException::class)
+    fun download(
+        formToDownload: ServerFormDetails,
+        files: List<MediaFile>,
+        tempMediaPath: String,
+        tempDir: File,
+        stateListener: OngoingWorkListener?
+    ) {
+        val tempMediaDir = File(tempMediaPath).also { it.mkdir() }
+
+        for (i in files.indices) {
+            stateListener?.progressUpdate(i + 1)
+
+            val mediaFile = files[i]
+            val tempMediaFile = File(tempMediaDir, mediaFile.filename)
+
+            val existingFile = searchForExistingMediaFile(formToDownload, mediaFile)
+            if (existingFile != null) {
+                copyFile(existingFile, tempMediaFile)
+            } else {
+                val mediaFile = formSource.fetchMediaFile(mediaFile.downloadUrl)
+                interuptablyWriteFile(mediaFile, tempMediaFile, tempDir, stateListener)
+            }
+        }
+    }
+
+    private fun searchForExistingMediaFile(
+        formToDownload: ServerFormDetails,
+        mediaFile: MediaFile
+    ): File? {
+        val allFormVersions = formsRepository.getAllByFormId(formToDownload.formId)
+        return allFormVersions.map { form: Form ->
+            File(form.formMediaPath, mediaFile.filename)
+        }.firstOrNull { file: File ->
+            val currentFileHash = getMd5Hash(file)
+            val downloadFileHash = validateHash(mediaFile.hash)
+            file.exists() && currentFileHash.contentEquals(downloadFileHash)
+        }
+    }
+
+    private fun validateHash(hash: String?): String? {
+        return if (hash == null || hash.isEmpty()) null else hash
```

Comment:
According to our implementation in `OpenRosaResponseParserImpl` media file hash can't be null or empty. If it is null or empty the whole file is skipped and not added to the list of media files. That means you should not need to validate it here.

Reply:
Hmmm good point. I don't know if we're ensuring that for media files when we parse everything. I'll look into that.

Reply:
Ok I dug into it and added a test to make it clearer what's happening here: https://github.com/getodk/collect/pull/5025/commits/bb5b5d00b98cb56c50177ce16fc916d42f666082. Basically you're right that we don't need to validate. The hash is never `null` as `MediaFile#hash` is of type `String` and if a media file has an empty hash the whole list will be returned as null which results in `OpenRosaFormSource` throwing an exception. It might be that we want to look at introducing some custom types like `NonEmptyString` to deal with this kind of thing, but I'm not sure if I want to do that now.